src/Symfony/Bundle/SecurityBundle/Form/Type/UsernamePasswordType.php

```diff
@@ -42,31 +41,23 @@ public function __construct(RequestStack $requestStack)
     public function buildForm(FormBuilderInterface $builder, array $options)
     {
         $builder
-            ->add('username', 'Symfony\Component\Form\Extension\Core\Type\TextType')
-            ->add('password', 'Symfony\Component\Form\Extension\Core\Type\PasswordType')
+            ->add('_username', 'Symfony\Component\Form\Extension\Core\Type\TextType')
+            ->add('_password', 'Symfony\Component\Form\Extension\Core\Type\PasswordType')
             ->add('_target_path', 'Symfony\Component\Form\Extension\Core\Type\HiddenType')
         ;
 
-        $request = $this->requestStack->getCurrentRequest();
-
         /* Note: since the Security component's form login listener intercepts
          * the POST request, this form will never really be bound to the
          * request; however, we can match the expected behavior by checking the
          * session for an authentication error and last username.
          */
-        $builder->addEventListener(FormEvents::PRE_SET_DATA, function (FormEvent $event) use ($request) {
-            if ($request->attributes->has(Security::AUTHENTICATION_ERROR)) {
-                $error = $request->attributes->get(Security::AUTHENTICATION_ERROR);
-            } else {
-                $error = $request->getSession()->get(Security::AUTHENTICATION_ERROR);
-            }
-
-            if ($error) {
+        $builder->addEventListener(FormEvents::PRE_SET_DATA, function (FormEvent $event) {
+            if (null !== $error = $this->authenticationUtils->getLastAuthenticationError()) {
                 $event->getForm()->addError(new FormError($error->getMessage()));
             }
 
             $event->setData(array_replace((array) $event->getData(), array(
-                'username' => $request->getSession()->get(Security::LAST_USERNAME),
+                '_username' => $this->authenticationUtils->getLastUsername(),
```

Comment:
@ro0NL : That's not exactly the same. Using `data` will lock the data and any event listener on `PRE_SET_DATA` would not be able to change it anymore.

https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Form.php#L323